### PR TITLE
#0: Port eltwise and some misc ops to use TensorSpec

### PR DESCRIPTION
**Pull request by sminakov-tt**
### Problem description
We need to migrate all ops to use `compute_output_specs` with TensorSpec, instead of older `compute_output_shapes`

### What's changed
Migrated ops to TensorSpec:
- binary
- unary
- bernoulli
- embedding
- embedding_backward
- unifrorm

Fixes in the infrastructure to support the migration.

### Checklist
- [x] [Post commit CI passes](https://github.com/tenstorrent/tt-metal/actions/runs/12040667994)
- [ ] Blackhole Post commit (if applicable)
- [ ] Model regression CI testing passes (if applicable)
- [ ] Device performance regression CI testing passes (if applicable)
- [x] New/Existing tests provide coverage for changes

### CODE REVIEW

File: ttnn/cpp/ttnn/operations/eltwise/binary/device/binary_device_operation.cpp [hunk 1]
```diff
@@ -150,11 +150,17 @@ void BinaryDeviceOperation::validate_on_program_cache_hit(
...
+    const auto& output_tensor = tensor_args.output_tensor;
+    if (output_tensor.has_value()) {
+        return output_tensor->get_tensor_spec();
+    }
+
+    const auto& input_tensor_a = tensor_args.input_tensor_a;
+    const auto input_shape_a = input_tensor_a.logical_shape();
     const auto& tensor_b = tensor_args.input_tensor_b;
-    const auto input_shape_b = tensor_b.has_value() ? tensor_b->shape() : ttnn::Shape{1, 1};
+    const auto input_shape_b = tensor_b.has_value() ? tensor_b->logical_shape() : ttnn::SimpleShape{};
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/eltwise/binary/device/binary_device_operation.cpp:163]: `Has value` what does it mean?

File: ttnn/cpp/ttnn/operations/eltwise/binary/device/binary_device_operation.cpp [hunk 2]
```diff
@@ -179,24 +185,9 @@ BinaryDeviceOperation::shape_return_value_t BinaryDeviceOperation::compute_outpu
...
-    auto output_shape = compute_output_shapes(operation_attributes, tensor_args);
-    const auto& input_tensor_a = tensor_args.input_tensor_a;
-    const auto& output_tensor = tensor_args.output_tensor;
-
-    if (output_tensor.has_value()) {
-        return output_tensor.value();
-    }
+    auto output_shape = compute_broadcasted_output(input_shape_a, input_shape_b);
 
     auto program_factory = select_program_factory(operation_attributes, tensor_args);
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/eltwise/binary/device/binary_device_operation.cpp:192]: Got a bit surprised that we call this method here. It means we call it at least two times per operation call. Not that it’s a problem but it’s a surprise.

File: ttnn/cpp/ttnn/operations/eltwise/binary/device/binary_device_operation.cpp [hunk 3]
```diff
@@ -224,16 +214,18 @@ BinaryDeviceOperation::tensor_return_value_t BinaryDeviceOperation::create_outpu
...
-        operation_attributes.memory_config);
+    return TensorSpec(output_shape, TensorLayout(operation_attributes.dtype, PageConfig(Layout::TILE), operation_attributes.memory_config));
+}
+
+BinaryDeviceOperation::tensor_return_value_t BinaryDeviceOperation::create_output_tensors(
+    const operation_attributes_t& operation_attributes, const tensor_args_t& tensor_args) {
+    if (tensor_args.output_tensor.has_value()) {
+        return *tensor_args.output_tensor;
+    }
+    return create_device_tensor(compute_output_specs(operation_attributes, tensor_args), tensor_args.input_tensor_a.device());
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/eltwise/binary/device/binary_device_operation.cpp:228]: So interesting. Tensor might have multiple devices today. I see this part of code did not change, but I wonder.. 

On this level multidevice basically does not exist today if I get it right.

**Review by ayerofieiev-tt — APPROVED** [review]
Please link test runs. 
Otherwise looks fine to me. 
Left some comments they would be good to discuss but they are not related to this change.

**Review by ntarafdar — APPROVED** [review]
I may have missed it but I don't see any padding calculations. Is this because none of these ops used the padded shape anywhere?

File: ttnn/cpp/ttnn/operations/eltwise/binary/device/binary_device_operation.cpp (hunk 1 above)
Comment (sminakov-tt) [reply ttnn/cpp/ttnn/operations/eltwise/binary/device/binary_device_operation.cpp:163]: `tensor_b` is optional here. I'm unsure what it means for a binary op to have a second argument as optional, but preserving the behavior

File: ttnn/cpp/ttnn/operations/eltwise/binary/device/binary_device_operation.cpp (hunk 2 above)
Comment (sminakov-tt) [reply ttnn/cpp/ttnn/operations/eltwise/binary/device/binary_device_operation.cpp:192]: Yeah. I would expect that `select_program_factory` should be really quick. It makes sense, at least to some extent, that the output tensor specs depend on the actual factory being used for these input arguments.

File: ttnn/cpp/ttnn/operations/eltwise/binary/device/binary_device_operation.cpp (hunk 3 above)
Comment (sminakov-tt) [reply ttnn/cpp/ttnn/operations/eltwise/binary/device/binary_device_operation.cpp:228]: Yes. Multi-device seems to be problematic overall at the moment. With having single MeshDevice it should work as expected though.

**Comment by sminakov-tt** [comment]
@ntarafdar None of the ops did anything before to request a custom padding.
In general, at this moment, ops shouldn't think too much about the padding, unless they want something non-standard.

**Comment by ntarafdar** [comment]
@sminakov-tt awesome. Ya next week I'm going to do the same for data movement ops. I think only transpose uses padding for anything so hoping it should be rather smooth for me as well!

File: ttnn/cpp/ttnn/operations/eltwise/binary/device/binary_device_operation.cpp (hunk 1 above)
Comment (ayerofieiev-tt) [reply ttnn/cpp/ttnn/operations/eltwise/binary/device/binary_device_operation.cpp:163]: I remember now. This is how float scalar is handled. Think of it like there is a union. It’s or float or tensor. Just expressed in this way

**Review by yan-zaretskiy — APPROVED** [review]

**Pull request merged**